diff --git a/quickjs.c b/quickjs.c
index 6774f84..8a6f45d 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -2280,6 +2280,11 @@ JSValue JS_GetClassProto(JSContext *ctx, JSClassID class_id)
     return js_dup(ctx->class_proto[class_id]);
 }
 
+JSValueConst JS_GetFunctionProto(JSContext *ctx)
+{
+    return ctx->function_proto;
+}
+
 typedef enum JSFreeModuleEnum {
     JS_FREE_MODULE_ALL,
     JS_FREE_MODULE_NOT_RESOLVED,
diff --git a/quickjs.h b/quickjs.h
index d449205..48078e6 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -323,6 +323,7 @@ JS_EXTERN void JS_SetContextOpaque(JSContext *ctx, void *opaque);
 JS_EXTERN JSRuntime *JS_GetRuntime(JSContext *ctx);
 JS_EXTERN void JS_SetClassProto(JSContext *ctx, JSClassID class_id, JSValue obj);
 JS_EXTERN JSValue JS_GetClassProto(JSContext *ctx, JSClassID class_id);
+JS_EXTERN JSValueConst JS_GetFunctionProto(JSContext *ctx);
 
 /* the following functions are used to select the intrinsic object to
    save memory */
