diff --git a/quickjs.c b/quickjs.c
index 48aeffc..fc145b8 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -6322,6 +6322,10 @@ JSValue JS_GetGlobalObject(JSContext *ctx)
     return JS_DupValue(ctx, ctx->global_obj);
 }
 
+JSValueConst JS_GetFunctionProto(JSContext *ctx) {
+    return ctx->function_proto;
+}
+
 /* WARNING: obj is freed */
 JSValue JS_Throw(JSContext *ctx, JSValue obj)
 {
diff --git a/quickjs.h b/quickjs.h
index d4a5cd3..c482686 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -785,6 +785,7 @@ JSValue JS_EvalThis(JSContext *ctx, JSValueConst this_obj,
                     const char *input, size_t input_len,
                     const char *filename, int eval_flags);
 JSValue JS_GetGlobalObject(JSContext *ctx);
+JSValueConst JS_GetFunctionProto(JSContext *ctx);
 int JS_IsInstanceOf(JSContext *ctx, JSValueConst val, JSValueConst obj);
 int JS_DefineProperty(JSContext *ctx, JSValueConst this_obj,
                       JSAtom prop, JSValueConst val,
