diff --git a/quickjs-libc.c b/quickjs-libc.c
index e180dd0..8219ae3 100644
--- a/quickjs-libc.c
+++ b/quickjs-libc.c
@@ -57,11 +57,6 @@ typedef sig_t sighandler_t;
 
 #endif
 
-#if !defined(_WIN32)
-/* enable the os.Worker API. IT relies on POSIX threads */
-#define USE_WORKER
-#endif
-
 #ifdef USE_WORKER
 #include <pthread.h>
 #include <stdatomic.h>
diff --git a/quickjs.c b/quickjs.c
index 48aeffc..cb6c80d 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -38,6 +38,8 @@
 #include <malloc.h>
 #elif defined(__FreeBSD__)
 #include <malloc_np.h>
+#else
+#include <malloc.h>
 #endif
 
 #include "cutils.h"
@@ -62,17 +64,6 @@
 #define MALLOC_OVERHEAD  8
 #endif
 
-#if !defined(_WIN32)
-/* define it if printf uses the RNDN rounding mode instead of RNDNA */
-#define CONFIG_PRINTF_RNDN
-#endif
-
-/* define to include Atomics.* operations which depend on the OS
-   threads */
-#if !defined(EMSCRIPTEN)
-#define CONFIG_ATOMICS
-#endif
-
 #if !defined(EMSCRIPTEN)
 /* enable stack limitation */
 #define CONFIG_STACK_CHECK
diff --git a/run-test262.c b/run-test262.c
index 2092cac..fc62eba 100644
--- a/run-test262.c
+++ b/run-test262.c
@@ -39,9 +39,6 @@
 #include "list.h"
 #include "quickjs-libc.h"
 
-/* enable test262 thread support to test SharedArrayBuffer and Atomics */
-#define CONFIG_AGENT
-
 #define CMD_NAME "run-test262"
 
 typedef struct namelist_t {
