patch = all 
diff --git a/quickjs.c b/quickjs.c
index 5ea8a12..7cce1dd 100644
--- a/quickjs.c
+++ b/quickjs.c
@@ -2108,6 +2108,18 @@ void JS_FreeRuntime(JSRuntime *rt)
     }
 }
 
+#ifdef CONFIG_PARALLEL
+void JS_ResetStackPointerRT(JSRuntime *rt)
+{
+     rt->stack_top = js_get_stack_pointer();
+}
+
+void JS_ResetStackPointer(JSContext *ctx)
+{
+     ctx->rt->stack_top = js_get_stack_pointer();
+}
+#endif
+
 JSContext *JS_NewContextRaw(JSRuntime *rt)
 {
     JSContext *ctx;
diff --git a/quickjs.h b/quickjs.h
index 4e06dfe..693f728 100644
--- a/quickjs.h
+++ b/quickjs.h
@@ -352,6 +352,12 @@ JSRuntime *JS_GetRuntime(JSContext *ctx);
 void JS_SetClassProto(JSContext *ctx, JSClassID class_id, JSValue obj);
 JSValue JS_GetClassProto(JSContext *ctx, JSClassID class_id);
 
+#ifdef CONFIG_PARALLEL
+void JS_ResetStackPointerRT(JSRuntime *rt);
+void JS_ResetStackPointer(JSContext *ctx);
+#endif
+
+
 /* the following functions are used to select the intrinsic object to
    save memory */
 JSContext *JS_NewContextRaw(JSRuntime *rt);
