--- quickjs/quickjs.h	2020-03-16 17:00:15.000000000 +0100
+++ quickjs/rquickjs.h	2020-04-10 16:54:24.285484917 +0200
@@ -351,6 +351,11 @@
 void JS_SetClassProto(JSContext *ctx, JSClassID class_id, JSValue obj);
 JSValue JS_GetClassProto(JSContext *ctx, JSClassID class_id);
 
+#ifdef CONFIG_PARALLEL
+void JS_ResetCtxStack(JSContext *ctx);
+#endif
+
+
 /* the following functions are used to select the intrinsic object to
    save memory */
 JSContext *JS_NewContextRaw(JSRuntime *rt);
@@ -997,6 +1002,13 @@
                        JSValue val);
 int JS_SetModuleExportList(JSContext *ctx, JSModuleDef *m,
                            const JSCFunctionListEntry *tab, int len);
+                           
+#ifdef CONFIG_MODULE_EXPORTS
+int JS_GetModuleExportEntriesCount(JSModuleDef *m);
+JSValue JS_GetModuleExportEntry(JSContext *ctx, JSModuleDef *m, int idx);
+JSAtom JS_GetModuleExportEntryName(JSContext *ctx, JSModuleDef *m, int idx);
+#endif
+
 
 #undef js_unlikely
 #undef js_force_inline
